gobblin-core/src/main/java/gobblin/writer/FsDataWriter.java

```diff
@@ -139,22 +139,33 @@ public FsDataWriter(FsDataWriterBuilder<?, D> builder, State properties) throws
     this.group = Optional.fromNullable(properties.getProp(ForkOperatorUtils
         .getPropertyNameForBranch(ConfigurationKeys.WRITER_GROUP_NAME, this.numBranches, this.branchId)));
 
-    if (this.group.isPresent()) {
-      HadoopUtils.setGroup(this.fs, this.stagingFile, this.group.get());
-    } else {
-      LOG.warn("No group found for " + this.stagingFile);
-    }
-
     // Create the parent directory of the output file if it does not exist
     WriterUtils.mkdirsWithRecursivePermission(this.fs, this.outputFile.getParent(), this.dirPermission);
   }
 
+  /**
+   * Create the staging output file and an {@link OutputStream} to write to the file.
+   *
+   * @return an {@link OutputStream} to write to the staging file
+   * @throws IOException if it fails to create the file and the {@link OutputStream}
+   */
   protected OutputStream createStagingFileOutputStream() throws IOException {
     return this.closer.register(this.fs.create(this.stagingFile, this.filePermission, true, this.bufferSize,
         this.replicationFactor, this.blockSize, null));
   }
 
   /**
+   * Set the group name of the staging output file.
+   *
+   * @throws IOException if it fails to set the group name
+   */
+  protected void setGroup() throws IOException {
```

Comment:
You can rename it to `setStagingFileGroup`. Also add a precondition check for the existence of the staging file.